crates/rooch-framework/sources/coin.move

```diff
@@ -0,0 +1,360 @@
+    public fun burn_from<CoinType>(
+        ctx: &mut StorageContext,
+        addr: address,
+        amount: u256,
+        burn_cap: &BurnCapability<CoinType>,
+    ) {
+        let coin_store = account_storage::global_borrow_mut<CoinStore<CoinType>>(ctx, addr);
+        let coin_to_burn = extract(&mut coin_store.coin, amount);
+        burn(ctx, coin_to_burn, burn_cap);
+    }
+
+    /// Destroys a zero-value coin. Calls will fail if the `value` in the passed-in `coin` is non-zero
+    /// so it is impossible to "burn" any non-zero amount of `Coin` without having
+    /// a `BurnCapability` for the specific `CoinType`.
+    public fun destroy_zero<CoinType>(zero_coin: Coin<CoinType>) {
+        let Coin { value } = zero_coin;
+        assert!(value == 0, error::invalid_argument(EDestroyOfNonZeroCoin))
+    }
+
+    /// Extracts `amount` from the passed-in `coin`, where the original coin is modified in place.
+    public fun extract<CoinType>(coin: &mut Coin<CoinType>, amount: u256): Coin<CoinType> {
+        assert!(coin.value >= amount, error::invalid_argument(EInSufficientBalance));
+        coin.value = coin.value - amount;
+        Coin { value: amount }
+    }
+
+    /// Extracts the entire amount from the passed-in `coin`, where the original coin is modified in place.
+    public fun extract_all<CoinType>(coin: &mut Coin<CoinType>): Coin<CoinType> {
+        let total_value = coin.value;
+        coin.value = 0;
+        Coin { value: total_value }
+    }
+
+    /// Freeze a CoinStore to prevent transfers
+    public fun freeze_coin_store<CoinType>(
+        ctx: &mut StorageContext,
+        addr: address,
+        _freeze_cap: &FreezeCapability<CoinType>,
+    ) {
+        let coin_store = account_storage::global_borrow_mut<CoinStore<CoinType>>(ctx, addr);
+        coin_store.frozen = true;
+    }
+
+    /// Unfreeze a CoinStore to allow transfers
+    public fun unfreeze_coin_store<CoinType>(
+        ctx: &mut StorageContext,
+        addr: address,
+        _freeze_cap: &FreezeCapability<CoinType>,
+    ) {
+        let coin_store = account_storage::global_borrow_mut<CoinStore<CoinType>>(ctx, addr);
+        coin_store.frozen = false;
+    }
+
+    /// Creates a new Coin with given `CoinType` and returns minting/freezing/burning capabilities.
+    /// The given signer also becomes the account hosting the information about the coin
+    /// (name, supply, etc.).
+    public fun initialize<CoinType>(
+        ctx: &mut StorageContext,
+        account: &signer,
+        name: string::String,
+        symbol: string::String,
+        decimals: u8,
+    ): (BurnCapability<CoinType>, FreezeCapability<CoinType>, MintCapability<CoinType>) {
+        let addr = signer::address_of(account);
+        assert!(
+            coin_address<CoinType>() == addr,
+            error::invalid_argument(ECoinInfoAddressMismatch),
+        );
+
+        assert!(
+            !account_storage::global_exists<CoinInfo<CoinType>>(ctx, addr),
```

Comment:
I prefer to save the CoinInfo to a global TypeTable, not in the coin's publisher's account storage.

Reply:
> I prefer to save the CoinInfo to a global TypeTable, not in the coin's publisher's account storage.

done